File: kernel/net/ip.c
/**
 * @file ip.c
 * @brief IP protocol implementation
 */

#include "../include/net/ip.h"
#include "../include/net/ethernet.h"
#include "../include/net/network.h"
#include "../include/net/arp.h"
#include "../include/kprintf.h"
#include "../include/debug.h"
#include "../include/mm/heap.h"
#include "../include/string.h"

/**
 * Calculate IP checksum
 */
uint16_t ip_checksum(ip_header_t* header) {
    uint32_t sum = 0;
    uint16_t* words = (uint16_t*)header;
    size_t header_len = (header->version_ihl & 0x0F) * 4;
    
    for (size_t i = 0; i < header_len / 2; i++) {
        sum += __builtin_bswap16(words[i]);
    }
    
    while (sum >> 16) {
        sum = (sum & 0xFFFF) + (sum >> 16);
    }
    
    return __builtin_bswap16(~sum);
}

/**
 * Send IP packet
 */
error_code_t ip_send(uint32_t dest_ip, uint8_t protocol, void* data, size_t len) {
    if (!data || len == 0) {
        return ERR_INVALID_ARG;
    }
    
    // Find a network device (use first available)
    net_device_t* device = network_find_device("eth0");
    if (!device || !device->up) {
        return ERR_DEVICE_NOT_FOUND;
    }
    
    // Allocate IP packet
    size_t packet_size = sizeof(ip_header_t) + len;
    ip_header_t* packet = (ip_header_t*)kmalloc(packet_size);
    if (!packet) {
        return ERR_OUT_OF_MEMORY;
    }
    
    // Build IP header
    packet->version_ihl = 0x45;  // IPv4, 5 words (20 bytes)
    packet->tos = 0;
    packet->total_length = __builtin_bswap16(packet_size);
    packet->identification = 0;
    packet->flags_fragment = 0;
    packet->ttl = 64;
    packet->protocol = protocol;
    packet->checksum = 0;
    packet->src_ip = device->ip_address;
    packet->dest_ip = dest_ip;
    
    // Copy data
    memcpy(packet->data, data, len);
    
    // Calculate checksum
    packet->checksum = ip_checksum(packet);
    
    // Determine destination MAC
    uint8_t dest_mac[6];
    uint32_t next_hop_ip;
    
    // Check if destination is broadcast
    if (dest_ip == 0xFFFFFFFF) {
        memset(dest_mac, 0xFF, 6);
    } else {
        // Check if destination is on same network
        uint32_t src_net = device->ip_address & device->netmask;
        uint32_t dest_net = dest_ip & device->netmask;
        
        if (src_net == dest_net) {
            // Same network
            next_hop_ip = dest_ip;
        } else {
            // Different network - send to gateway
            next_hop_ip = device->gateway;
        }
        
        // Resolve MAC address via ARP
        error_code_t err = arp_resolve(next_hop_ip, dest_mac);
        if (err != ERR_OK) {
            kfree(packet);
            kerror("IP: ARP resolution failed for %08x\n", next_hop_ip);
            return err;
        }
    }
    
    // Send via Ethernet
    error_code_t err = ethernet_send(packet, packet_size, dest_mac, ETH_TYPE_IPV4, device);
    
    kfree(packet);
    return err;
}

/**
 * Receive IP packet
 */
error_code_t ip_receive(net_device_t* device, void* buffer, size_t len) {
    if (!device || !buffer) {
        return ERR_INVALID_ARG;
    }
    
    // Cast buffer to IP header (Ethernet header already stripped or we point to payload)
    // ethernet.c passes pointer to payload
    ip_header_t* ip_header = (ip_header_t*)buffer;
    
    // Verify checksum
    uint16_t checksum = ip_header->checksum;
    ip_header->checksum = 0;
    uint16_t calculated = ip_checksum(ip_header);
    ip_header->checksum = checksum;
    
    if (checksum != calculated) {
        return ERR_FAILED;  // Invalid checksum
    }
    
    // Check if packet is for us
    if (ip_header->dest_ip != device->ip_address) {
        // Check for broadcast
        if (ip_header->dest_ip != 0xFFFFFFFF) {
            return ERR_NOT_FOUND;  // Not for us
        }
    }
    
    // Extract payload
    size_t header_len = (ip_header->version_ihl & 0x0F) * 4;
    size_t payload_len = __builtin_bswap16(ip_header->total_length) - header_len;
    
    uint8_t proto = ip_header->protocol;
    uint32_t src = ip_header->src_ip;
    
    void* payload = (uint8_t*)buffer + header_len;
    
    // Handle ICMP packets
    if (proto == IP_PROTOCOL_ICMP) {
        extern error_code_t icmp_handle_packet(void* buffer, size_t len, uint32_t src_ip);
        icmp_handle_packet(payload, payload_len, src);
    }
    
    return ERR_OK;
}

File: services/network/src/ip.rs
//! IP protocol implementation

use core::mem;

/// IP header structure
#[repr(C, packed)]
pub struct IpHeader {
    pub version_ihl: u8,      // Version (4 bits) + IHL (4 bits)
    pub tos: u8,              // Type of Service
    pub total_length: u16,    // Total length
    pub identification: u16,  // Identification
    pub flags_fragment: u16,  // Flags (3 bits) + Fragment offset (13 bits)
    pub ttl: u8,              // Time to Live
    pub protocol: u8,         // Protocol
    pub checksum: u16,        // Header checksum
    pub src_ip: u32,          // Source IP address
    pub dst_ip: u32,          // Destination IP address
    pub data: [u8; 0],        // Variable length data
}

/// IP protocol numbers
pub const IP_PROTOCOL_ICMP: u8 = 1;
pub const IP_PROTOCOL_TCP: u8 = 6;
pub const IP_PROTOCOL_UDP: u8 = 17;

/// Calculate IP checksum
pub fn ip_checksum(header: &IpHeader) -> u16 {
    let mut sum: u32 = 0;
    let header_len = ((header.version_ihl & 0x0F) * 4) as usize;
    let words = unsafe {
        core::slice::from_raw_parts(header as *const _ as *const u16, header_len / 2)
    };
    
    for &word in words {
        sum += u16::from_be_bytes(word.to_le_bytes()) as u32;
    }
    
    while (sum >> 16) != 0 {
        sum = (sum & 0xFFFF) + (sum >> 16);
    }
    
    !(sum as u16).to_be()
}

/// Send IP packet
pub fn ip_send(dest_ip: u32, protocol: u8, data: &[u8]) -> Result<(), ()> {
    // Get network device via IPC
    // For now, use first available device
    // Allocate packet buffer
    let mut packet = [0u8; 1500];
    
    // Build IP header
    let mut ip_header = IpHeader {
        version_ihl: 0x45, // IPv4, 5 * 4 = 20 bytes header
        tos: 0,
        total_length: (20 + data.len()) as u16,
        identification: 0,
        flags_fragment: 0,
        ttl: 64,
        protocol,
        checksum: 0,
        src_ip: 0, // Would get from network device
        dst_ip: dest_ip,
    };
    
    // Calculate checksum
    ip_header.checksum = ip_checksum(&ip_header);
    
    // Copy header and data to packet
    unsafe {
        core::ptr::copy_nonoverlapping(&ip_header as *const _ as *const u8, packet.as_mut_ptr(), 20);
    }
    let data_len = data.len().min(1480);
    packet[20..20+data_len].copy_from_slice(&data[0..data_len]);
    
    // Send via Ethernet
    use crate::ethernet_device::send_packet;
    let _ = send_packet(&packet[0..20+data_len]);
    
    Ok(())
}

/// Receive IP packet
pub fn ip_receive(buffer: &mut [u8]) -> Result<(usize, u32, u8), ()> {
    // Receive from Ethernet layer
    use crate::ethernet_device::receive_packet;
    let mut eth_buffer = [0u8; 1518];
    match receive_packet(&mut eth_buffer) {
        Ok(len) => {
            if len < 20 {
                return Err(());
            }
            // Parse IP header
            let ip_header = unsafe {
                &*(eth_buffer.as_ptr().add(14) as *const IpHeader) // Skip 14-byte Ethernet header
            };
            
            // Verify IP version
            if (ip_header.version_ihl >> 4) != 4 {
                return Err(());
            }
            
            // Extract data
            let header_len = ((ip_header.version_ihl & 0x0F) * 4) as usize;
            let data_len = (ip_header.total_length as usize).saturating_sub(header_len);
            let copy_len = data_len.min(buffer.len());
            buffer[0..copy_len].copy_from_slice(&eth_buffer[14+header_len..14+header_len+copy_len]);
            
            // Return data length, source IP, protocol
            Ok((copy_len, ip_header.src_ip, ip_header.protocol))
        }
        Err(_) => Err(())
    }
}

